### ISSUE TITLE
"Missing return statement" error for functions having infinite loop unconditionally by looping on non-exhaustive generator

### ISSUE THREAD

**Issue by yxliang01** [python/mypy#7374]
When having functions having infinite loop unconditionally by looping on non-exhaustive generator, there should be no `return` statements after the infinite looping block since all statements after the looping block are unreachable. Therefore, this "Missing return statement" error is a false-positive. The test case exposing this false-positive is implemented in https://github.com/python/mypy/pull/7373 .

Thanks

**Comment by JelleZijlstra** [comment]
The type system does not know whether a generator is finite (or exhaustive as you call it), so there's no way for mypy to know this. We could add a new type system feature, like an "infinite iterator" type, but I'm not sure that's worth the hassle.

I would recommend putting something like `assert False, "unreachable"` in functions affected by this bug, which also makes it clear to human readers why there appears to be a missing return statement.

**Comment by yxliang01** [comment]
Yes. The mitigation is what I'm currently doing. But, I think that for built-in ones, at least white-listing is a possibility?

**Comment by ilevkivskyi** [comment]
This is actually a duplicate of https://github.com/python/mypy/issues/5992 that was closed as wont fix. So I would propose to close this as well, unless you want to provide a working patch.

**Comment by ilevkivskyi** [comment]
It if fine to keep this open, but it I don't think it is high priority. PRs are welcome however.

**Comment by yxliang01** [comment]
I am not familiar with mypy codebase and, unfortunately, my schedule doesn't allow me to master this codebase and do this... But, thanks for keeping this issue. How should we do about #7373 ?

**Comment by ilevkivskyi** [comment]
Honestly, I would just close that PR for now. Then later we can re-open it after this issue is fixed.

**Comment by sbrudenell** [comment]
Reading the typing module docs, I want to believe that infinite generators could be annotated as `Generator[X, Y, NoReturn]`.

Is that a stupid idea? Or just stupid enough to work? :smile:

**Comment by gvanrossum** [comment]
@sbrudenell Why don't you try it and report back here?

**Comment by sbrudenell** [comment]
```python
from typing import Generator
from typing import NoReturn

def inf() -> Generator[None, None, NoReturn]:
    while True:
        yield

def consume_inf() -> int:
    for _ in inf():
        return 0

def consume_while_true() -> int:
    while True:
        return 0
```

```bash
$ mypy --version
mypy 0.790
$ mypy test.py
test.py:8: error: Missing return statement
Found 1 error in 1 file (checked 1 source file)
```

I'd expect `consume_inf()` to type-check, as `consume_while_true()` does. Today it complains only about `consume_inf()`.

(In real usage, my `inf()`s are often something like [this comment](https://github.com/python/mypy/issues/5992#issuecomment-610828566), especially for writing unit tests to wait for some condition with a timeout)

I'm trying to work out whether `Generator[X, Y, NoReturn]` is a good idea.

Pro:
* It's easy to use, in the use cases I can find. One could make it even more readable with something like `InfiniteIterator = Generator[_T, None, NoReturn]`
* I can't think of use cases it would interfere with (actually, I can't figure out what the `ReturnType` of `Generator` is ever used for)
* It's intuitive from the perspective of writing a generator function: the function should never return
* It's intuitive from the perspective of loops: the loop never exits on its own, just like a `while True`

Con:
* It's unintuitive / an impedence mismatch from the perspective of the low-level generator API: a constraint on `ReturnType` really means that `.__next__()`, `.send()` and `.throw()` can't raise `StopIteration`. Constraining exceptions is a new idea in the type system
  * (however, I have not seen the generator API widely used, except for implementing event loops. If mypy only naively supported this change, in the context of `for` and `while` control flow, I think it would cover a huge majority of the use cases)

I can't quite work out how to change mypy to implement this for further testing. I guess there ought to be some change around `TypeChecker.accept_loop()` or its callers. But I haven't figured out what the change should be, and I've run out of patience to shave this particular yak.

**Comment by sbrudenell** [comment]
Adding to the Cons:
* I originally thought `Generator[X, Y, NoReturn]` wasn't a _new feature_, just adding support for `NoReturn` in a context it wasn't supported before. But the return type of `Generator` isn't as simple as that.
